src/java/org/apache/cassandra/service/reads/AbstractReadExecutor.java

```diff
@@ -225,17 +229,27 @@ public boolean hasLocalRead()
      */
     boolean shouldSpeculateAndMaybeWait()
     {
+        long now = preciseTime.now();
+        long sampleLatencyNanos = MICROSECONDS.toNanos(cfs.sampleReadLatencyMicros);
         // no latency information, or we're overloaded
-        if (cfs.sampleReadLatencyMicros > command.getTimeout(MICROSECONDS))
+        if (sampleLatencyNanos > command.getTimeout(NANOSECONDS))
         {
-            if (logger.isTraceEnabled())
-                logger.trace("Decided not to speculate as {} > {}", cfs.sampleReadLatencyMicros, command.getTimeout(MICROSECONDS));
+            logger.trace("Decided not to speculate as {}ns > {}ns", sampleLatencyNanos, command.getTimeout(NANOSECONDS));
             return false;
         }
 
+        if (now + sampleLatencyNanos > requestTime.clientDeadline())
```

Comment:
Is the `enqueuedAtNanos` that informs `clientDeadline()` also a precise time?

Reply:
Yes. `enqueuedAtNanos` is precise: 

in `Request` constructor:
```
createdAtNanos = MonotonicClock.Global.preciseTime.now();
```

and then

```
processRequest(channel, request, forFlusher, backpressure, new RequestTime(request.createdAtNanos, startTimeNanos));
```

Also, all metrics are using precise time.